azblob/zc_retry_reader.go

```diff
@@ -53,14 +54,32 @@ type retryReader struct {
 	countWasBounded bool
 	o               RetryReaderOptions
 	getter          HTTPGetter
+
+	// forced re-read handling (which is the only thread-safe part of this type)
+	rrLock      *sync.Mutex
```

Comment:
Do we absolutely need a lock? 

I assume the main purpose is to protect `forceRetry()`, which must set `rrForced` and also cancel the HTTP Body at the same time. Then `wasForcedRetry()` could return the right value. But the lock doesn't seem to solve the race condition completely, because what if an error that shouldn't be retried is returned from `s.response.Body.Read(p)`,  and then `rrForced` is set to true; by the time we get to line 121 and call `wasForcedRetry()`, we'd end up retrying an error that shouldn't be retried.

Reply:
Yes, the use in `forceRetry()` is its main purpose.  And the lock avoids the scenario where `wasForcedRetry()` returns false even "after" the HTTP body has been cancelled.  So the lock correctly protects us from thinking we were _not_ cancelled when really we _were_.  That protection is important because, without it, we could occasionally misinterpret forced retries as a non-retryable error, and fail the transfer.  I.e. we fail a transfer that should _not_ have failed.

As you say, the lock doesn't protect us from the opposite situation, i.e.:  thinking we _were_  cancelled when really we _weren't_.  But IMHO the consequences of that are less severe than failing the transfer. The consequence is just that we retry the GET/read sequence.  Presumably that retry would fail for the same fatal reason, and then we'd (correctly) identify it as non-retryable. I.e. we fail a transfer that _should_ have failed. We just take a little longer to do it.

To solve the latter situation the only option I could think of was to check err in `wasForcedRetry` to see if it looks like a cancellation error.  But even that is not technically perfect, and the context cancelled error looks like something that we'd have to check for by its string message/content, which seemed a bit ugly.  Any suggestions a on better way to solve the latter issue, and how badly we need to do so?

Reply:
We agreed, in verbal discussion, that is good enough as is.